src/main/java/org/gridsuite/shortcircuit/server/computation/service/AbstractWorkerService.java

```diff
@@ -0,0 +1,240 @@
+                    publishFail(resultContext, e.getMessage());
+                    resultService.delete(resultContext.getResultUuid());
+                }
+            } finally {
+                futures.remove(resultContext.getResultUuid());
+                cancelComputationRequests.remove(resultContext.getResultUuid());
+            }
+        };
+    }
+
+    public Consumer<Message<String>> consumeCancel() {
+        return message -> cancelAsync(CancelContext.fromMessage(message));
+    }
+
+    protected abstract void saveResult(Network network, AbstractResultContext<R> resultContext, S result);
+
+    protected abstract void sendResultMessage(AbstractResultContext<R> resultContext, S result);
+
+    protected abstract void publishFail(AbstractResultContext<R> resultContext, String message);
+
+    /**
+     * Do some extra task before running the computation, e.g. print log or init extra data for the run context
+     * @param ignoredRunContext This context may be used for further computation in overriding classes
+     */
+    protected void preRun(R ignoredRunContext) {
+        LOGGER.info("Run {} computation ...", getComputationType());
+    }
+
+    protected S run(Network network, R runContext, UUID resultUuid) throws Exception {
+        String provider = runContext.getProvider();
+        AtomicReference<Reporter> rootReporter = new AtomicReference<>(Reporter.NO_OP);
+        Reporter reporter = Reporter.NO_OP;
+
+        if (runContext.getReportInfos() != null && runContext.getReportInfos().reportUuid() != null) {
+            final String reportType = runContext.getReportInfos().computationType();
+            String rootReporterId = runContext.getReportInfos().reporterId() == null ? reportType : runContext.getReportInfos().reporterId() + "@" + reportType;
+            rootReporter.set(new ReporterModel(rootReporterId, rootReporterId));
+            reporter = rootReporter.get().createSubReporter(reportType, String.format("%s (%s)", reportType, provider), "providerToUse", provider);
+            // Delete any previous computation logs
+            observer.observe("report.delete",
+                    runContext, () -> reportService.deleteReport(runContext.getReportInfos().reportUuid(), reportType));
+        }
+        runContext.setReporter(reporter);
+
+        preRun(runContext);
+        CompletableFuture<S> future = runAsync(network, runContext, provider, resultUuid);
+        S result = future == null ? null : observer.observeRun("run", runContext, future::get);
+        postRun(runContext);
+
+        if (runContext.getReportInfos().reportUuid() != null) {
+            for (final AbstractReportMapper reportMapper : reportMappers) {
+                rootReporter.set(reportMapper.processReporter(rootReporter.get()));
+            }
+            observer.observe("report.send", runContext, () -> reportService.sendReport(runContext.getReportInfos().reportUuid(), rootReporter.get()));
```

Comment:
we can move these implementations into the overriden method postRun() but we need to pass rootReporter as one of its arguments

Reply:
In the last version the reporter is inside the runContext.

Reply:
Yes this `AbstractReportMapper` seems quite specific to ShortCircuit. If there is no plan to use it for the other microservices it shouldn't be included in the generic classes. 

Maybe ask to whoever created it ? @Tristan-WorkGH ?